### ISSUE TITLE
Method references for Optional trigger raw type false positive

### ISSUE THREAD

**Issue by andrewparmet** [google/error-prone#694]
Thanks for a great tool! I found this bit of funniness today.

Error Prone's compiler warns about missing type arguments for the following method:

```java
static long foo(Stream<String> stream, Function<String, ? extends Optional<Integer>> fn) {

    // Does not trigger
    long a = stream.map(fn)
    .filter(opt -> opt.isPresent())
    .count();

    // Triggers Error Prone
    long b = stream.map(fn)
    .filter(Optional::isPresent)
    .count();

    return a + b;
}
```
```
        .filter(Optional::isPresent)
                ^
  missing type arguments for generic class Optional<T>
```

As far as I can see this is a false positive.  The method reference line does not trigger Error Prone when `fn`'s second type parameter is `Optional<Integer>` and not `? extends Optional<Integer>`.

**Comment by andrewparmet** [comment]
I see this appears to be a duplicate of https://github.com/google/error-prone/issues/569.

**Comment by jbduncan** [comment]
@andrewparmet I'm not so sure this is a dupe of #569. To me, it looks like #569 is about enums, whereas this issue is about non-enum classes like `Optional`. Thus I'd think this issue is sufficiently different and worth reopening.

**Comment by andrewparmet** [comment]
I thought this was more of an issue regarding instance methods on generic classes (such as Optional and Enum), but I'll reopen and see if anyone else has anything to say.

**Comment by cushon** [comment]
Sorry, I linked to the wrong OpenJDK bug in #569. This is the same issue, thanks for reporting.
